### ISSUE TITLE
Media query API

### ISSUE THREAD

**Issue by mattjbray** [rtfeldman/elm-css#307]
The type-safe media queries are a cool idea :)

However I think the generated CSS is incorrect.

For reference, here's the BNF (from [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/Media_Queries/Using_media_queries#Pseudo-BNF), also a less readable version in the [CSS3 spec](https://drafts.csswg.org/mediaqueries-3/#syntax)):

```
media_query: [[only | not]? <media_type> [ and <expression> ]*]
           | <expression> [ and <expression> ]*
expression: ( <media_feature> [: <value>]? )
```

(I'm not an expert on media queries -- I've only read the linked MDN page -- so if I have this all wrong please ignore.)

## Case 1: Single expression

This snippet:

```elm
css =
    Css.stylesheet
        [ Media.media (Media.minWidth (px 768))
            [ body [ color (hex "f00") ] ]
        ]
```

generates:

```css
@media only (min-width: 768px) {
    body {
        color: #f00;
    }
}
```

which is incorrect because `only` may only appear before a `<media_type>`.

## Case 2: Specifying a media type

This snippet:

```elm
css =
    Css.stylesheet
        [ Media.media (Media.and Media.screen (Media.minWidth (px 768)))
            [ body [ color (hex "f00") ] ]
        ]
```

generates:

```css
@media only (screen and (min-width: 768px)) {
    body {
        color: #f00;
    }
}
```

which is incorrect because `(screen and ...)` is parenthesised - only individual expressions should be in parentheses.

## Case 3: Invalid specifications

It's also possible to specify invalid media queries in elm-css.

For example:

* Specifying two media types with `and` is not valid:

   ```elm
   Media.media (Media.and Media.screen Media.print)
   ```

   generates
   
   ```css
   @media only (screen and print)
   ```

* `Media.or` assumes an `or` keyword that doesn't exist (queries should be comma-separated instead):

   ```elm
   Media.media (Media.or Media.screen Media.print)
   ```
   
   generates
   
   ```css
   @media only (screen or print)
   ```

* Only top-level queries may be negated, but elm-css allows negating individual or nested expressions:

   ```elm
   media (and screen (not (minWidth (px 768))))
   ```

   generates

   ```css
   @media only (screen and not (min-width: 768px))
   ```

## Proposal

In the spirit of making impossible states impossible, the last case motivates a new API for media queries. I think this representation also makes the CSS code generation rules more straightforward.

```elm
type MediaQuery
type MediaType
type Expression

-- Media.media takes a List of queries (there is no Media.or).
media : List MediaQuery -> List Snippet -> Snippet

-- Create a MediaQuery that applies to all MediaTypes.
-- Media.all (Media.minWidth (px 768))
-- => /* @media (minWidth: 768px) */
all : Expression -> MediaQuery

-- Combining Expressions
-- Media.all (Media.and (Media.minWidth (px 768)) (Media.maxWidth (px 992)))
-- => /* @media (minWidth: 768px) and (maxWith: 992px) */
and : Expression -> Expression -> Expression

-- Specify a MediaType
-- Media.only Media.print
-- => /* @media only print */
only : MediaType -> MediaQuery

-- Media type with expression(s)
-- Media.only2 Media.print (Media.and Media.color (Media.orientation Media.portrait))
-- => /* @media only print and (color) and (orientation: portrait) */
only2 : MediaType -> Expression -> MediaQuery

-- Negate a media type
-- Media.not Media.screen
-- => /* @media not screen */
not : MediaType -> MediaQuery
not2 : MediaType -> Expression -> MediaQuery


-- Media types
screen : MediaType
...

-- Expressions
minWidth : AbsoluteLength compatible -> Expression
...
```

What do you think?

**Comment by rtfeldman** [comment]
Nice finds @mattjbray! I really appreciate your taking the time to dig into this and write this up.

I think your proposed API is great overall. It makes a ton of sense to me!

What do you think of this variation?

```elm
all : Expression -> MediaQuery
only : MediaType -> List Expression -> MediaQuery
not : MediaType -> List Expression -> MediaQuery
```

The idea would be that instead of having `only2`and `not2`, you take a `List Expression` and join the expressions together with `and`. If you provide an empty list, then you get no expressions (same as `not` and `only` in your original proposal).

Unless I'm missing something (likely), it would appear this also lets us get rid of `and`.

Thoughts?

**Comment by rtfeldman** [comment]
Another thought - looking at this, I realize that `Css.Media.MediaQuery` should probably be `Css.Media.Query` - so the type would be `Query` instead of `MediaQuery`. Seem reasonable?

**Comment by mattjbray** [comment]
Thanks!

> The idea would be that instead of having `only2` and `not2`, you take a `List Expression`

Absolutely. I actually started with this idea but thought you might want to stick closer to CSS syntax with `Media.and`. I prefer your suggestion, it's much simpler.

> Maybe it's better if they're `MediaQuery` and `MediaType` after all.

I prefer this as it's the terminology used by the spec.

So I'm in agreement with your proposal in https://github.com/rtfeldman/elm-css/issues/307#issuecomment-325222806 (plus the change to `media` to accept a `List MediaQuery`.

**Comment by rtfeldman** [comment]
@mattjbray Awesome! Want to take a crack at a PR?

**Comment by mattjbray** [comment]
@rtfeldman Sure thing. I have a long train journey tomorrow and this will fill the time nicely :)

**Comment by rtfeldman** [comment]
Fixed by #308
